docs/_docs/03-ug-mutable-dataset.md

```diff
@@ -87,7 +89,84 @@ hs.refreshIndex("empIndex", "incremental")
 ```
 
 ### Optimize Index
-TODO
+The fastest way to index newly added source data files and merge them into an existing index is by calling `refreshIndex`
+with the `"incremental"` mode. In this mode, each time refresh is called, it creates fresh index files for new data files
+and updates index metadata to include them in the index content.
+As these index files accumulate, they could affect query performance when the index is used.
+Once the index is leveraged for a query,
+the large number of these files could increase overall query time as more index files need to be accessed and potentially
+read to compute the query results.
+Hyperspace provides the `optimizeIndex` command to alleviate above issue by changing index files layout for an
+index which has many index files, due to incremental refresh call(s). This is achieved by merging index files together,
+if possible, and replacing them with fewer larger files that capture exact same index records. This process is similar to
+compaction in append-only log structured merge index structures.
+
+You should note that the `optimizeIndex` command is a best effort to modify index files layout and its
+final outcome depends on how index records are stored in existing index files.
+Two or more index files can be merged with each other, if and only if they all have index records which belong to the
+same bucket (according to index configuration).
+When running `optimizeIndex` command, Hyperspace tries to find such groups of index files and merge them together.
+However, if most of the index files have records from different buckets, then final physical layout of index files
```

Comment:
> if most of the index files have records from different buckets, 

Could you rephrase this part?

Reply:
I made a bit of change here. Please let me know if that addressed your comment.